### DISCUSSION TITLE
Add OBS websocket support through "Vendor" requests and event

### DISCUSSION THREAD

**Discussion by tainalo2** [sorayuki/obs-multi-rtmp#244] [Ideas]
Hello,

Thank you Sorayuki for your work !
I'm working on a web-app to control OBS from smartphone when you are AFK (like IRL streaming).
It's working with OBS websocket 5.0.
I would like to control multi-rtmp plugin this way too, to simply add or modify output parameters.
Since 5.0 OBS websocket support "Vendor" request to use it with third party plugins (see in next link).

Is it possible to implement it in multi-rtmp please ?

OBS WS vendor documentation : https://github.com/obsproject/obs-websocket/blob/master/docs/generated/protocol.md#callvendorrequest

**Comment by sorayuki** [comment]
It seems a very useful function. Thank you for informing me. I will take a look later.

**Comment by siddolo** [comment]
hi @sorayuki any news about websocket support?

Thanks

**Comment by gdantuono** [comment]
hi @sorayuki, I would like to start, stop the stream via websocket and get the statistics

**Comment by mbakumenkov** [comment]
It would be a very cool update! @sorayuki please, take a look :)

**Comment by Aeon1704** [comment]
It would have been great to have. Apps like Streamer.bot will have easier access and able to read bitrate data, connection errors and stuff. Thus, streamers will be able to make notifications scripts and so on, without need to learn programming languages to make C# requests.

**Comment by Aeon1704** [comment]
Any news on this? Or maybe example of how to get plugin's bitrate datas over C# method? Streamer bot for example have a C# code sub action interacting with OBS using Websocket.
Basic code:
```
using System;

public class CPHInline
{
	public bool Execute()
	{
		// your main code goes here
		return true;
	}
}
```
But, since plugin doesn't have any documentation, websocket vendor or otherwise, i don't really want to try 'guess'.

**Comment by davidcool** [comment]
I downloaded the source and was able to get web socket server access to most items working. I was able to get it compiled for PC and I have several Python code examples for access and control via OBS Web Socket. I'm happy to share though it's definitely a fork at this point. Does anyone have any guidance or expertise at compiling for MacOS? I've been hacking at it but with no success yet. If anyone wants the source, or just the compiled plugin for PC let me know. I can also provide the Python examples as well.

**Reply by Aeon1704** [reply]
> I downloaded the source and was able to get web socket server access to most items working. I was able to get it compiled for PC and I have several Python code examples for access and control via OBS Web Socket. I'm happy to share though it's definitely a fork at this point. Does anyone have any guidance or expertise at compiling for MacOS? I've been hacking at it but with no success yet. If anyone wants the source, or just the compiled plugin for PC let me know. I can also provide the Python examples as well.

Well, i am on Windows, it could be cool to have Websocket and Streamer bot control like: Read bitrate, stop and start via script. Atleast you prove that it can be done. And honestly, at this point, official OBS devs should just buy MultiRTMP plugin and include it in main build, and support it in future :D

**Reply by davidcool** [reply]
I just release a version of obs-multi-rtmp with websocket API, but I could only figure out how to compile for Windows. You can check it out here: https://github.com/davidcool/obs-multi-rtmp/releases/tag/v0.7.3.0-ws0.1

Let me know if you try it out!

**Reply by Aeon1704** [reply]
Does it have independent connection? Or where can i find to json request example?

**Reply by davidcool** [reply]
If you download the package, inside is a "python_tests" folder that has Python examples of connecting to the OBS websocket server and modifying or returning most core features. In your post you mentioned wanting to return the stats for a particular target. I've attached one of the python examples here for you to look at, but all of the other examples are in the download package.
[8_test_stats.py.zip](https://github.com/user-attachments/files/22559195/8_test_stats.py.zip)
I only provided Python examples, but you could use whatever language you'd like to interface with the websocket server and return the results.

**Reply by Aeon1704** [reply]
Thanks, figured it out, Streamer.bot actually getting info from main Websocket.

**Reply by Aeon1704** [reply]
But it seems i was unable to use Streamer.bot OBS-Raw request with Target Name to get stats. So i've put "name": "Youtube" - didn't work. But "id": "numbers" - did work.

**Reply by davidcool** [reply]
Yeah, you have to use the ID to return details about a target, name isn't available. You can return the target name via ListTargets though.

**Reply by Aeon1704** [reply]
You really should try and push it to Sorayuki, so it will stay as main code in repository. Because that kind of feature will help alot of streamers. Me for example using Streamer.bot to control actions on OBS with all forms of logics and variables. very same Streamer bot will be able to perform start \ stops for Multi-RTMP with Websocket support, and warn about faulty connection.

**Reply by davidcool** [reply]
Yeah, I'd love if Sorayuki integrated it into the main plugin as well. I did mention it, but I'm not sure if it's the direction they want to take the plugin or not, we'll see. I will try and maintain parallel releases with the added websocket support until it's integrated into the main, or if sorayuki isn't interested in that direction, I'll try and keep releasing the forked versions. If you know anyone who can help compile this for MacOS or Linux, do let me know! Also, if you run into bugs, please submit them on my repository so I can keep track of them, but only bugs pertaining to the websocket side of things. https://github.com/davidcool/obs-multi-rtmp-websocket-support

**Comment by Aeon1704** [comment]
After using Davidcool's version with Websocket support, it is very hard now to give up. Streamer bot app works nicely with OBS and Multi-RTMP Websocket. My Streamer Bot scripts are able to Read data and warn me about faulty connection or bitrate\fps drops on certain RTMP outputs. Aswell as carefully start streams one by one, via the very same Streamer Bot scripting. Please consider to add Websocket into your official build :)
